clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -1380,6 +1384,9 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
 
   ModulePassManager MPM;
 
+  if (CodeGenOpts.DisableLLVMPasses && LangOpts.SYCLIsDevice)
+    MPM.addPass(PropagateAspectUsagePass());
```

Comment:
```suggestion
  if (LangOpts.SYCLIsDevice)
    MPM.addPass(PropagateAspectUsagePass());
```

Reply:
`DisableLLVMPasses` check is needed here or otherwise we will add the pass twice. However, I would prefer if we only had a single insertion point for this pass instead of two like it is suggested in this PR

Reply:
In case `DisableLLVMPasses == false` we should add this pass in `registerPipelineStartEPCallback`, otherwise we should add just with `addPass` method. This detail is the reason why there are 2 insertion points.

Reply:
It is not necessary add this pass via `registerPipelineStartEPCallback` in general. I proposed such solution when there were no case for `DisableLLVMPasses == true`.
Taking into account a requirement to add this pass even for `DisableLLVMPasses == false` case it is better to add it once on line 1388.

Reply:
Ok, @maksimsab helped me to figure out that we should add `PropagateAspectUsagePass` in two places because later in `if (!CodeGenOpts.DisableLLVMPasses)` branch `ModulePassManager` instance is recreated and all that is added via `addPass` method before that moment is lost. On the other hand, functions which really add passes that are included via `registerPipeline*Callback` methods also called inside that branch.